hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -151,23 +135,33 @@ private ExecutionPlanBuilder() {
         return list;
     }
 
-    public static Map<Address, int[]> getPartitionAssignment(NodeEngine nodeEngine) {
-        final Address[] partitionOwners = new Address[nodeEngine.getPartitionService().getPartitionCount()];
-        initPartitionOwnersAndMembers(nodeEngine, partitionOwners);
+    /**
+     * Assign the partitions to their owners. Partitions whose owner isn't in
+     * the {@code memberList}, are assigned to one of the members in a
+     * round-robin way.
+     */
+    public static Map<MemberInfo, int[]> getPartitionAssignment(NodeEngine nodeEngine, List<MemberInfo> memberList) {
+        IPartitionService partitionService = nodeEngine.getPartitionService();
+        Map<Address, MemberInfo> membersByAddress = memberList.stream().collect(toMap(MemberInfo::getAddress, identity()));
+
+        final MemberInfo[] partitionOwners = new MemberInfo[partitionService.getPartitionCount()];
+        int memberIndex = 0;
+        for (int partitionId = 0; partitionId < partitionOwners.length; partitionId++) {
+            Address address = partitionService.getPartitionOwnerOrWait(partitionId);
+            MemberInfo member = membersByAddress.get(address);
+            if (member == null) {
+                // if the partition owner isn't in the current memberList, assign to one of the other members in
+                // round-robin fashion
+                member = memberList.get(memberIndex++ % memberList.size());
+            }
```

Comment:
It seems to me that this partition assignment made here may cause logic errors (at least in any wrong use in the processors). If possible, could you explain this partition assignment in mixed cluster in the design doc? I don't know how a job reading the map or a scan query would behave in a mixed cluster?

Reply:
Fixed, see https://github.com/hazelcast/hazelcast/pull/19026/commits/bd79951743603c8442635c16acaeca725ca32eae